### ISSUE TITLE
rootof: Constructing RootOfs with polys containing RootOf coefficients

### ISSUE THREAD

**Issue by cswiercz** [sympy/sympy#9366]
This is a feature request to make it possible to construct `RootOf`s containing `RootOf`s like so:

```python
from sympy import Poly, RootOf
from sympy.abc import x,y

aroot = RootOf(x**7 + x + 1, 0, radicals=False)
p = RootOf(y**2 - aroot, 0, radicals=False)
```

I did a little investigating of my own to see what is preventing the current implementation from supporting this. Here's what I've done so far. It's incredible naive but since I'm unfamiliar with the internals of Sympy domains it's the best I can do at the moment. I would be happy to do what I can to make this work. It would be a humungous benefit to my own project!

1. When running the code above we get a `NotImplementedError` during construction of the `RootOf` object. I didn't see where in [`RootOf.__new__`](https://github.com/sympy/sympy/blob/master/sympy/polys/rootoftools.py#L54) why this is an issue so I started by commenting out [Lines 97-98](https://github.com/sympy/sympy/blob/master/sympy/polys/rootoftools.py#L97) from `rootoftools.py` just to see what would happen.

  ```
  if not dom.is_ZZ:                                                                                                                                              
      raise NotImplementedError("RootOf is not supported over %s" % dom)                                                                                         
  ```

2. Executing the example again brought us to a `DomainError` while trying to compute the isolating intervals of the root. In particular, [`RootOf._get_reals_sqf`](https://github.com/sympy/sympy/blob/master/sympy/polys/rootoftools.py#L152) makes a call to [`rootofisolation.dup_isolate_real_roots_sqf`](https://github.com/sympy/sympy/blob/master/sympy/polys/rootisolation.py#L490) with the following arguments:
  * `f = [EX(1), EX(0), EX(-RootOf(x**7 + x + 1, 0))]`
  * `K = EX`
  * `eps, inf, sup = None`,
  * `basis = False`
  * `fast = True`

  The error is raised on [Lines 503-504](https://github.com/sympy/sympy/blob/master/sympy/polys/rootisolation.py#L503) because `EX` is not `ZZ`. See [Trace 1] below for the stack trace. So, continuing this line of inquiry I commented out those lines.

3. What follows next is beyond my immediate understanding. I suggest taking at look at [Trace 2] below but from what little I understand it looks like an error occurs in trying to coerce some input values to [`_mobius_to_interval`](https://github.com/sympy/sympy/blob/master/sympy/polys/rootisolation.py#L144) to the field `EX`. In this case, the inputs to this function are
  * `M = (EX(0), EX(1), EX(1), EX(1))`
  * `field = EX`

  The error occurs when trying to execute `EX(M[0],M[2])`. A full trace is below at [Trace 2].

At this point I stopped because I don't know enough about the inner workings of domains to make any decent judgements. If anyone would like to play around I [created a branch](https://github.com/cswiercz/sympy/tree/rootofrootof) with the above two sections commented out. Again, I would be happy to work with someone more qualified to get this working!

[Trace 1] After First Commenting
--------------------------------------------

```
---------------------------------------------------------------------------
DomainError                               Traceback (most recent call last)
<ipython-input-4-cc81d1e3e947> in <module>()
      1 get_ipython().magic(u'pdb 0')
      2 
----> 3 p = RootOf(y**2 - r, 0, radicals=False)
      4 print p

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in __new__(cls, f, x, index, radicals, expand)
     98         #     raise NotImplementedError("RootOf is not supported over %s" % dom)
     99 
--> 100         root = cls._indexed_root(poly, index)
    101         return coeff*cls._postprocess_root(root, radicals)
    102 

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in _indexed_root(cls, poly, index)
    406         (_, factors) = poly.factor_list()
    407 
--> 408         reals = cls._get_reals(factors)
    409         reals_count = cls._count_roots(reals)
    410 

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in _get_reals(cls, factors)
    178 
    179         for factor, k in factors:
--> 180             real_part = cls._get_reals_sqf(factor)
    181             reals.extend([ (root, factor, k) for root in real_part ])
    182 

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in _get_reals_sqf(cls, factor)
    157             _reals_cache[factor] = real_part = \
    158                 dup_isolate_real_roots_sqf(
--> 159                     factor.rep.rep, factor.rep.dom, blackbox=True)
    160 
    161         return real_part

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootisolation.pyc in dup_isolate_real_roots_sqf(f, K, eps, inf, sup, fast, blackbox)
    502         (_, f), K = dup_clear_denoms(f, K, convert=True), K.get_ring()
    503     elif not K.is_ZZ:
--> 504         raise DomainError("isolation of real roots not supported over %s" % K)
    505 
    506     if dup_degree(f) <= 0:

DomainError: isolation of real roots not supported over EX
```

[Trace 2] After Second Commenting
-------------------------------------------------

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-f3b6aafd55bc> in <module>()
      1 get_ipython().magic(u'pdb 1')
      2 
----> 3 p = RootOf(y**2 - r, 0, radicals=False)
      4 print p

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in __new__(cls, f, x, index, radicals, expand)
     98         #     raise NotImplementedError("RootOf is not supported over %s" % dom)
     99 
--> 100         root = cls._indexed_root(poly, index)
    101         return coeff*cls._postprocess_root(root, radicals)
    102 

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in _indexed_root(cls, poly, index)
    406         (_, factors) = poly.factor_list()
    407 
--> 408         reals = cls._get_reals(factors)
    409         reals_count = cls._count_roots(reals)
    410 

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in _get_reals(cls, factors)
    178 
    179         for factor, k in factors:
--> 180             real_part = cls._get_reals_sqf(factor)
    181             reals.extend([ (root, factor, k) for root in real_part ])
    182 

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootoftools.pyc in _get_reals_sqf(cls, factor)
    157             _reals_cache[factor] = real_part = \
    158                 dup_isolate_real_roots_sqf(
--> 159                     factor.rep.rep, factor.rep.dom, blackbox=True)
    160 
    161         return real_part

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootisolation.pyc in dup_isolate_real_roots_sqf(f, K, eps, inf, sup, fast, blackbox)
    509     I_zero, f = _isolate_zero(f, K, inf, sup, basis=False, sqf=True)
    510 
--> 511     I_neg = dup_inner_isolate_negative_roots(f, K, eps=eps, inf=inf, sup=sup, fast=fast)
    512     I_pos = dup_inner_isolate_positive_roots(f, K, eps=eps, inf=inf, sup=sup, fast=fast)
    513 

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootisolation.pyc in dup_inner_isolate_negative_roots(f, K, inf, sup, eps, fast, mobius)
    463     elif not mobius:
    464         for f, M in roots:
--> 465             u, v = _mobius_to_interval(M, F)
    466             results.append((-v, -u))
    467     else:

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/rootisolation.pyc in _mobius_to_interval(M, field)
    146     a, b, c, d = M
    147 
--> 148     s, t = field(a, c), field(b, d)
    149 
    150     if s <= t:

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/domains/domain.pyc in __call__(self, *args)
     70     def __call__(self, *args):
     71         """Construct an element of ``self`` domain from ``args``. """
---> 72         return self.new(*args)
     73 
     74     def normal(self, *args):

/Users/cswiercz/.local/lib/python2.7/site-packages/sympy-0.7.7.dev0-py2.7.egg/sympy/polys/domains/domain.pyc in new(self, *args)
     62 
     63     def new(self, *args):
---> 64         return self.dtype(*args)
     65 
     66     @property

TypeError: __init__() takes exactly 2 arguments (3 given)
```

**Comment by jksuom** [comment]
The implementation of the ``RootOf`` class is currently restricted to complex roots of polynomials with rational (or integral) coefficients. The roots are identified and labelled  by isolating them into disjoint intervals, real or complex. The isolation code is based on the ordering of real numbers. The biggest obstacle to extending the current implementation to polynomials such as ``y**2 - aroot`` is the lack of order in their coefficient fields. In this case, ``aroot`` is probably the unique negative real root of ``x**7 + x + 1``, so ``QQ(aroot)`` would be a real algebraic number field, but the order induced by that of the reals is currently not implemented in such fields.

To sum up: If you need to isolate the roots in the complex field, the first step would be the implementation of order in real algebraic number fields. The case of a non-real coefficient field would be harder to implement.

**Comment by cswiercz** [comment]
> To sum up: If you need to isolate the roots in the complex field, the first step would be the implementation of order in real algebraic number fields. The case of a non-real coefficient field would be harder to implement.

It sounds like the greater issue is not that algebraic extension fields are not ordered but that there isn't a root isolation algorithm for such fields implemented in Sympy. The technique used with polynomials over `QQ`uses interval arithmetic which, of course, depends on the fact that `RR` is totally ordered. Has anyone in the Sympy community worked on this problem?

Since a numerical approximation is used to identify `RootOf`s anyway can we use that approximation in constructing the "second level `RootOf`s"? That is, perhaps something like the following can be done to obtain those isolating intervals:

```python
from sympy import Poly, RootOf, sqrt
from sympy.abc import x,y

aroot = RootOf(x**7 + x + 1, 0, radicals=False)
p = RootOf(y**2 - aroot.n(), 0, radicals=False)   # use numerical approximation

print p
print p.n()
print sqrt(aroot.n()).n()   # turns out to be the other square root
```
```
RootOf(17592186044416*y**2 + 14012956470457, 0)
-0.892493335621313*I
0.892493335621313*I
```

The roots of a polynomial are continuous as a function of their coefficients so, since numerical approximations are used anyway, using an approximate coefficient will get us intervals which will be close enough.

**Comment by cswiercz** [comment]
Actually, this won't work when `aroot.n()` is strictly complex.

```python
aroot = RootOf(x**7 + x + 1, 1, radicals=False)  # root #1 has a non-zero imaginary part
p = RootOf(y**2 - aroot.n(), 0, radicals=False)
```
```
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
...
NotImplementedError: RootOf is not supported over EX
```

So a different algorithm is needed.

**Comment by cswiercz** [comment]
Something similar is available in Sage. The [`complex_roots`](https://github.com/sagemath/sage/blob/master/src/sage/rings/polynomial/complex_roots.py#L250) function returns complex roots of type `ComplexIntervalFieldElement` which, at least in terms of representing a `RootOf` via a numerical approximation, is similar to what's happening in Sympy. However, the algorithm they use to refine the intervals seems...heuristic, based on a large comment at the top of [`refine_root`](https://github.com/sagemath/sage/blob/master/src/sage/rings/polynomial/complex_roots.py#L35) that begins with:

> There has got to be a better way to do this, but I don't know
   what it is...

:) Anyway, here's a sample session performing the calculation above in Sage.

```python
from sage.rings.polynomial.complex_roots import complex_roots

x = polygen(ZZ)
p = x**7 + x + 1
proots = complex_roots(p, retval='algebraic')
​print 'proots:'
print proots

aroot = roots[0][0]
y = polygen(ZZ)
q = y**2 + aroot

roots = complex_roots(q)
print '\nqroots:'
print qroots
```

```
proots:
[(-0.7965443541284571?, 1), (-0.7052980879371503? - 0.6376237698062315?*I, 1), (-0.7052980879371503? + 0.6376237698062315?*I, 1), (0.12376188051147740? - 1.056650011197074?*I, 1), (0.12376188051147740? + 1.056650011197074?*I, 1), (0.9798083844899014? - 0.5166768838399121?*I, 1), (0.9798083844899014? + 0.5166768838399121?*I, 1)]

qroots:
[(-0.892493335621313?, 1), (0.892493335621313?, 1)]
```

I had to change the types of `proots` to elements of `QQbar` in order to construct the polynomial `q`. However, this is not an issue in Sympy with `RootOf`s. Anyway, I guess we could try to implement a similar algorithm to the one in `refine_interval` if something better isn't known by someone else in the community.

**Comment by jksuom** [comment]
I know of no one who had worked with these problems recently. If you really need the roots mapped into the complex field, the following might be a workable alternative. 

If ``K`` is an algebraic number field and ``p`` is a polynomial in ``K[X]``, you can form the *norm* ``Np`` of ``p`` over ``Q``, i.e., the product of the conjugates of ``p``, It has rational coefficients, so its roots can be isolated with the current algorithm.  All the roots of ``p`` are among those of ``Np``, you only have to identify them.

The high degree of ``Np`` may become a problem, but, on the other hand, working with non-rational coefficient fields is also computationally intensive.

**Comment by cswiercz** [comment]
@jksuom Thank you for your response!

Is there a particular place in the `RootOf` "pipeline" where the use of the number field norm would be inserted? From what I understand somewhere in `polys/rootisolation.py` would be good, maybe in `dup_refine_real_root`?  Any suggestions on where to start writing code would be greatly appreciated.

**Comment by jksuom** [comment]
> where the use of the number field norm would be inserted?

I don't think it would be possible to simply insert the norm into the isolation code. The norm of the polynomial should be constructed  before the root isolation (i.e. ``RootOf``)  is attempted.

If the field ``K`` has ``t`` as a primitive element, then your polynomial can be written in the form ``p(t, X)``. If ``q(t)`` is the minimal polynomial of ``t``, then the norm is the resultant of ``p`` and ``q`` with respect to ``t``. (Essentially the elimination of ``t``.)

**Comment by cswiercz** [comment]
@jksuom I stepped away from this problem for a bit due to thesis work. Now I'm back. Thank you for the responses above.

The following is mostly for myself as I think about how to implement this on my own (and want to document my thoughts), but it turns out that Sympy can already determine the norm of `p(t,X)`:

```python
from sympy import Dummy, RootOf, Poly
from sympy.abc import x
_x = Dummy(str(x))
t = RootOf(_x**5 + 4*_x + 2, 0)
p = Poly(x**4 + t*x + 1, x, extension=t) # or, using "domain=QQ.algebraic_field(s)"

print p
print p.domain.is_Algebraic
print p.sqf_norm()
```
```
Poly(x**4 - RootOf(_x**5 + 4*_x + 2, 0)*x + 1, x, domain='QQ<-RootOf(_x**5 + 4*_x + 2, 0)>')
True
(0, Poly(x**4 - RootOf(_x**5 + 4*_x + 2, 0)*x + 1, x, domain='QQ<-RootOf(_x**5 + 4*_x + 2, 0)>'), Poly(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, x, domain='QQ'))
```

So that's nice. (Although, I tried it for the example `f = (x**4 + t*x + 1)**2`, let it run for approximately 30 minutes on my machine, and still hadn't received a result.)

Now, I need to figure out how to use this norm with the complex root isolation algorithm. From what I understand, the algorithm is based on the [Collins and Krandick](http://dl.acm.org/citation.cfm?id=143308) paper. Perhaps there's a way to modify it...

**Comment by jksuom** [comment]
@cswiercz ``sqf_norm`` is related to the norm we a looking for but not necessarily the same. Recall that the norm is the product of all conjugates of the polynomial ``p(t, X)``, i.e. of the polynomials ``p(t', X)`` where ``t'`` runs through all conjugates of ``t``. If these conjugates have common roots (thus making the product non-squarefree), ``sqf_norm`` replaces the polynomial with ``p(t, X - st)`` where ``s`` is a constant (a small integer). The idea is that if ``p(t, X)``  and ``p(t', X )`` have a common root ``x``, then the corresponding roots ``x + st`` of ``p(t, X - st)`` and ``x + st'`` of ``p(t', X - st')`` will no longer be the same. But then the result of ``sqf_norm`` is not the norm of ``p(t, X)`` if ``s != 0``,

If the original polynomial ``p(t, X)`` has multiple roots then all the modified polynomials ``p(t, X - st)`` also have multiple roots and ``sqf_norm`` will never succeed. (That explains the failure with your ``f``.) So it should be only applied to squarefree polynomials, and even then it may fail to give the correct norm. To find that one has to take the resultant. (That is what ``sqf_norm`` actually does, for each tested value of ``s``.)

**Comment by cswiercz** [comment]
@jksuom Since the norm is multiplicative would taking the product of the `sqf_norm`s of each term in the square-free factorization of `p` (along with appropriate exponents) be sufficient? (Or, simply compute the roots of each square-free factor of `p` and collect.) If not, I can implement the algorithm described in [Algebraic Factoring and Rational Function Integration](http://delivery.acm.org/10.1145/810000/806338/p219-trager.pdf?ip=69.91.149.223&id=806338&acc=ACTIVE%20SERVICE&key=B63ACEF81C6334F5%2EF43F328D6C8418D0%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&CFID=674757849&CFTOKEN=53093662&__acm__=1433351000_902723d64e23ffdd13c434aed2738a0b) by Trager to compute the norm of `p`. (Which is based on a similar technique using resultants.)

For simplicity, assume `p(t,x)` is square-free and `q(x)` is the minimal polynomial for `t`. Then, the roots `alpha` of the resultant `R(x) = Res_y(p(y,x), q(y))` are such that the equations `p(y,alpha) = 0` and `q(y)=0` have simultaneous solutions. The solutions are exactly `y=t'` where `t'` runs through all conjugates of the original algebraic number `t`.

But each of these `alpha` work for the specific choice `y=t`. Therefore, the roots of `p(t,x) = 0` are `alpha = RootOf(R(x), i)` for `i=0,...,deg(R)-1`. But then it seems like some additional "testing" needs to be done since `deg_x(R)` is typically much larger than `deg_x(P)`.

I wrote the following to test the above algorithm sketch. 

```python
from sympy import Dummy, RootOf, Poly, resultant
from sympy.abc import x,y

_x = Dummy(str(x))
t = RootOf(_x**5 + 4*_x + 2, 0)
p = Poly(x**4 + t*x + 1, x, extension=t) # or, using "domain=QQ.algebraic_field(s)"
print 'p      =', p.as_expr()

# we want to compute the resultant R(x) = Res_y(p(x,y),q(y))
# where q is the minpoly of t
q = t.poly.eval(y)
py = p.as_expr().xreplace({t:y})
r = resultant(py,q,y)
print 'q(y)   =', q.as_expr()
print 'p(x,y) =', py.as_expr()
print 'r(x)   =', r

# determine the roots of the resultant
print '\nComputing and evaluating roots:' 
alpha = r.as_poly(x).all_roots(radicals=False)
for alpha_i in alpha:
    print '\nalpha_i =', alpha_i
    print 'p(alpha_i,t) =', p.as_expr().subs(x,alpha_i).n()
```
```
p      = x**4 - x*RootOf(_x**5 + 4*_x + 2, 0) + 1
q(y)   = y**5 + 4*y + 2
p(x,y) = x**4 - x*y + 1
r(x)   = x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1

Computing and evaluating roots:

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 0)
p(alpha_i,t) = -0.198524332438419 + 1.22554688478738*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 1)
p(alpha_i,t) = -0.198524332438419 - 1.22554688478738*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 2)
p(alpha_i,t) = -2.41944778909615 - 0.566701491125257*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 3)
p(alpha_i,t) = -2.41944778909615 + 0.566701491125257*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 4)
p(alpha_i,t) = 0.e-131 - 0.e-132*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 5)
p(alpha_i,t) = 0.e-131 + 0.e-132*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 6)
p(alpha_i,t) = 0.201916462283013 - 2.16764838738598*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 7)
p(alpha_i,t) = 0.201916462283013 + 2.16764838738598*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 8)
p(alpha_i,t) = 0.637726119575724 - 0.262890650102781*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 9)
p(alpha_i,t) = 0.637726119575724 + 0.262890650102781*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 10)
p(alpha_i,t) = 1.09212546803052 - 0.212543988319327*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 11)
p(alpha_i,t) = 1.09212546803052 + 0.212543988319327*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 12)
p(alpha_i,t) = 0.820925000684375 + 0.864933480581428*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 13)
p(alpha_i,t) = 0.820925000684375 - 0.864933480581428*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 14)
p(alpha_i,t) = 0.e-129 - 0.e-131*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 15)
p(alpha_i,t) = 0.e-129 + 0.e-131*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 16)
p(alpha_i,t) = -1.26012678782168 - 0.623504054308736*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 17)
p(alpha_i,t) = -1.26012678782168 + 0.623504054308736*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 18)
p(alpha_i,t) = 1.12540585878262 - 1.81349088458005*I

alpha_i = RootOf(x**20 + 5*x**16 + 10*x**12 + 14*x**8 + 2*x**5 + 9*x**4 + 1, 19)
p(alpha_i,t) = 1.12540585878262 + 1.81349088458005*I
```

As expected, there are four x-roots `alpha_i = RootOf(R(x),i)` for `i=4,5,14,15` of the polynomial `p(x,t)`. n

Do you know if there is a computationally efficient way to isolate these four roots out of the twenty possible ones without having to go through all this work? Also, I can't really think of where this would fit in the `RootOf` constructor if, say, I wanted to do something like

```python
t = RootOf(_x**5 + 4*_x + 2, 0)
r = RootOf(_y**4 + t*_y + 1)
```
I guess a separate subroutine would have to be written for polynomials with non-integer coefficients.

**Comment by jksuom** [comment]
> Or, simply compute the roots of each square-free factor of p and collect.

That is exactly what I had in mind. (That should be done to keep the degrees as low as possible, anyway.)

> But then it seems like some additional "testing" needs to be done since deg_x(R) is typically much larger than deg_x(P).

That is what I tried to point out briefly above.  ("All the roots of p are among those of Np, you only have to identify them.")

> Do you know if there is a computationally efficient way to isolate these four roots out of the twenty possible ones without having to go through all this work?

I do not know of any other method than scanning through the list of all roots of the norm even though that may feel inefficient. It might be possible to generalize the root isolation algorithm for algebraic base fields, but that would probably also be computationally less efficient.

**Comment by cswiercz** [comment]
> That is what I tried to point out briefly above. ("All the roots of p are among those of Np, you only have to identify them.")

Well, that's all I need, I think, to write up a pull request. Thank you very much for your input!

I'll, of course, link to this discussion when the PR is created.

**Comment by skirpichev** [comment]
> Is there a particular place in the RootOf "pipeline" where the use of the number field norm would be inserted?

@cswiercz,  perhaps, you can do this in the class constructor.  Also, you should modify ``RootOf``'s class methods, like ``all_roots()``.  See POC PR in my project, diofant/diofant#430.

Unfortunately, root isolation is too slow in SymPy (or Diofant) for polys of high degree.

@jksuom, do you think that such canonicalization of ``RootOf``'s, to have polynomials with integer coefficients, does make sense per default (like Mathematica does)?

**Comment by jksuom** [comment]
@skirpichev By 'like Mathematica does', do you refer to this: `Root[{f1,f2,…},{k1,k2,…}]`?

BTW, I don't consider passing to the norm of a polynomial as an ideal solution. The degree of the norm is high and only few of its roots are also roots of the initial polynomial.

**Comment by skirpichev** [comment]
> @skirpichev By 'like Mathematica does', do you refer to this: Root[{f1,f2,…},{k1,k2,…}]?

@jksuom no, this: ``Root[f,k] is automatically reduced so that f has the smallest possible degree and smallest integer coefficients. ``

For example:
```
In[2]:= Root[
 Function[x, (7 + 3 Sqrt[2]) x^3 + (-6 - 4 Sqrt[2]) x^2 + (-Sqrt[2] - 
      1) x + 2], 1]

Out[2]= Root[
 4 - 4 #1 - 25 #1^2 + 24 #1^3 + 2 #1^4 - 36 #1^5 + 31 #1^6 &, 2]
```

> The degree of the norm is high and only few of its roots are also roots of the initial polynomial.

Yes, but it's doable in principle (as Mathematica shows).  Currently implemented algorithm for root isolation [could be improved](https://mattpap.github.io/masters-thesis/html/src/algorithms.html#complex-roots).

**Comment by jksuom** [comment]
Always passing over to integer (or rational) coefficients does not seem a good idea to me. I would rather extend the root isolation algorithm for polynomials with real algebraic coefficients. That should be possible because they can be ordered compatibly with real numbers.

**Comment by skirpichev** [comment]
> extend the root isolation algorithm for polynomials with real algebraic coefficients. 

But real algebraic coefficients - only a special case of algebraic coefficients.  And, unless I misinterpret your opinion:

>  It might be possible to generalize the root isolation algorithm for algebraic base fields, but that would probably also be computationally less efficient.

- for complex algebraic numbers some generalization of the root isolation algorithm will be much more costly.

**Comment by jksuom** [comment]
I forget the details but I probably had in mind the possibility separating real and imaginary parts of the coefficients in the same way as is currently done with the powers of the unknown variable. Extra processing would be needed, but maybe the total cost would still be of the same order of magnitude.

**Comment by skirpichev** [comment]
Ok, root isolation algorithms now supports gaussian rationals in the Diofant.  It seems, that computations slightly faster if coefficients are kept algebraic:
```python
In [4]: p = x**7 + I*x**4 + 2*x - I

In [5]: %time RootOf(p, x, 0, extension=True)
CPU times: user 17.8 s, sys: 20 ms, total: 17.8 s
Wall time: 17.8 s
Out[5]: 
      ⎛ 14      8      4      2       ⎞
RootOf⎝x   + 5⋅x  - 2⋅x  + 4⋅x  + 1, 0⎠

In [6]: %time RootOf(p, x, 0)
CPU times: user 2.29 s, sys: 0 ns, total: 2.29 s
Wall time: 2.3 s
Out[6]: 
      ⎛ 7      4             ⎞
RootOf⎝x  + ⅈ⋅x  + 2⋅x - ⅈ, 0⎠
In [7]: _5.n(3)
Out[7]: -1.1 - 0.68⋅ⅈ

In [8]: _6.n(3)
Out[8]: -1.1 - 0.68⋅ⅈ
```

Probably, algorithms could be extended for arbitrary algebraic coefficients.  For this we must extend ``dup_real_imag`` to support algebraic numbers beyond gaussian rationals and implement comparisions ops for real number fields.  No idea how badly the degree of extension will affects computation.

**Comment by cswiercz** [comment]
Wow! That's fantastic that you posted progress on this problem after two years!

**Comment by skirpichev** [comment]
Real algebraic coefficients also not poses any problem (see draft implementation in diofant/diofant#/630).

> I forget the details but I probably had in mind the possibility separating real and imaginary parts of the coefficients in the same way as is currently done with the powers of the unknown variable.

@jksuom, I think that real/imag parts are defined well only for CM-fields.  Is there some algorithm to produce such extensions of QQ for polynomials with algebraic coefficients?  Or you are keeping in mind some other possibility?

**Comment by jksuom** [comment]
I think that you already have an algorithm for (an isolation interval of) the real part of an algebraic number. It should be possible to do the same for the imaginary part. Assuming that each algebraic coefficient can be written as `c = a + ib` where `a` and `b` are real it should be possible to separate the real and imaginary parts of any polynomial with algebraic coefficients in `z = x + iy`. For example, `cz^2` would become `(ax^2 - 2bxy - ay^2) + i(bx^2 +2axy - by^2)`.

**Comment by skirpichev** [comment]
> I think that you already have an algorithm for (an isolation interval of) the real part of an algebraic number.

For finding real roots of the polynomial with real algebraic coefficients.

> It should be possible to do the same for the imaginary part.

The tricky part is: when notion of real and imaginary parts does make sense.

Lets we have some set of algebraic numbers (coefficients of the given polynomial, say ``{c_k}).  The problem is how to construct a 1) suitable extension of QQ, that has these numbers and that taking real/imaginary parts in this extension does make sense. And 2) another extension of QQ, real, that has all these real/imaginary parts.

Probably, the following should work.  Lets write ``c_k = a_k + I*b_k``, then make 2) extension, say ``A``, using the list of all real/imaginary coefficients, ``A = QQ.algebraic_field(a_1, ..., b_1, ...)``, make 1) extension by adding imaginary unit: ``B = A.algebraic_field(I)``.

**Comment by jksuom** [comment]
If you begin with the minimal polynomial (over Q) of an algebraic number, then its real part is a root of [this resultant](https://github.com/diofant/diofant/blob/master/diofant/polys/rootisolation.py#L1814). A polynomial with rational coefficients having the imaginary part as a root could also be constructed as another resultant. That way it should be possible to get the real and imaginary parts as real algebraic numbers.

I would start by finding the real and imaginary parts of each algebraic coefficient `c_k`, and then convert them to elements of a common (big) real number field. That would then be the coefficient field of the two polynomials (re and im) used to isolate the roots of the polynomial with the coefficients `c_k`.

**Comment by skirpichev** [comment]
> I would start by finding the real and imaginary parts of each algebraic coefficient c_k, and then convert them to elements of a common (big) real number field. That would then be the coefficient field of the two polynomials (re and im) used to isolate the roots of the polynomial with the coefficients c_k.

I think, this is same construction as I told above.  I.e. we construct first a real field ``A``, which contains all re/im parts of coefficients ``c_k``.  Then construct another extension ``B = A.algebraic_field(I)``, which contains all ``c_k``, ``I`` and conjugated ``c_k`` as well.  We can take real/imaginary parts in this field much like in the field of gaussian rationals.  Is that correct?

If so, unfortunately, AlgebraicField.domain could be only QQ in sympy/diofant.  Do you think this class must be extended first to support arbitrary tower of fields?

**Comment by jksuom** [comment]
> we construct first a real field A, which contains all re/im parts of coefficients c_k. 

I think that this would suffice for root isolation. The two polynomials with real algebraic coefficients could be constructed at this point.

> Then construct another extension B = A.algebraic_field(I), which contains all c_k, I and conjugated c_k as well. We can take real/imaginary parts in this field much like in the field of gaussian rationals.

I think that this would not be necessary for root isolation. But if we want to have a field with fast complex conjugation, then a construction of type `Complex(A)` (field of pairs of elements of `A`)  might be better. It is not needed for the conjugates of `c_k` as they are already contained in `QQ.algebraic_field(c_k)`. (Of course, `B = A.algebraic_field(I)` could be useful otherwise.)

> this class must be extended first to support arbitrary tower of fields

It might be handy to consider extensions like `QQ -> A -> B` with two number fields but I would hesitate to introduce arbitrary towers.

**Comment by skirpichev** [comment]
> I would hesitate to introduce arbitrary towers.

Hmm, why?  Given the fact we already have something like this for transcendental extensions, i.e. ``QQ.frac_field(x).frac_field(y).frac_field(z)``.

**Comment by jksuom** [comment]
I meant towers that would have the whole hierarchy coded in a single class. That is, a sequence of "domains" instead of a single `domain`. But I think that the domain of an algebraic field could be another algebraic field. Then towers could be constructed by chaining extensions as in case of `QQ.frac_field(x).frac_field(y).frac_field(z)`.

**Comment by skirpichev** [comment]
> But I think that the domain of an algebraic field could be another algebraic field. Then towers could be constructed by chaining extensions

Yes, I also meant this option.  I think it shouldn't be too hard: ``minimal_polynomial()`` does work in the Diofant with algebraic fields, the missing thing is support ``domain`` option also in the ``primitive_element()``.

> then a construction of type Complex(A) (field of pairs of elements of A) might be better

Such subclass of ``Field`` also could be useful.  It might cover gaussian rationals, root isolation needs and currect  ComplexField for inexact complex numbers.  I'm not sure about naming, just ``Complex`` doesn't feel right.
